fix(channel-web): database indices

BPMJoannette opened the pull request:
Adding indices to various tables and reworked a select query for improved overall performance. With migration file to update existing installations.

BPMJoannette committed:
fix(core): database performance

BPMJoannette committed:
removed hanging line

BPMJoannette committed:
removed foreign key reference

BPMJoannette committed:
fix(core): database performance

allardy commented:
The index on web_messages should be on conversationId AND userId, otherwise it will not be useful since all users will share a similar conversation ID. What is the impact of excluding "visit" messages ? Won't it still do a full table scan to get messages ? 

For web_conversations, we also use the field `last_heard_on` in most conditions, would there be an improvement by adding it? 

Just wondering, have you ever had an error when adding indexes? A bit curious for the message "Could not add indices"

This ran fine on Postgres, but didn't test sqlite yet. Did you run some performance tests to see what kind of gain we get from this ?

BPMJoannette commented:
> The index on web_messages should be on conversationId AND userId, otherwise it will not be useful since all users will share a similar conversation ID. What is the impact of excluding "visit" messages ? Won't it still do a full table scan to get messages ?

The query made on web_messages references the conversationId only; the userId is checked against the web_conversations table, therefore the userId is unnecessary in the web_messages index. Excluding visit messages is also mandatory on retrieving the conversation messages, so this maps directly to the condition. Testing shows that Postgres prefers this index to making a sequential scan.

> For web_conversations, we also use the field `last_heard_on` in most conditions, would there be an improvement by adding it?

The slow query is from listConversations, which orders by `last_heard_on`, but no condition on it there, so it is unnecessary to index as well. We do have a condition on it in the getOrCreateRecentConversation that could benefit from it, though it didn't come up as a slow query (or at all).

> Just wondering, have you ever had an error when adding indexes? A bit curious for the message "Could not add indices"

Didn't get any errors, and using the CREATE IF NOT EXISTS should handle most cases, but added the try catch anyway for good measure, for example if a table or column in these suddenly stops existing for some reason.

> This ran fine on Postgres, but didn't test sqlite yet. Did you run some performance tests to see what kind of gain we get from this ?

I tested SQLite as well, both work well, though I did not test the performance on SQLite.
On a 1GB web_messages table, I went from over 2 seconds to 40 ms for the conversations query, so that amounts to about 97% better performance for big tables.

Hope this addresses most of your concerns!

EFF committed:
Merge branch 'master' into mj-perfdb

allardy reviewed (approved):
Tested quickly on postgres and sqlite, update went well
